src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs

```diff
@@ -495,6 +495,131 @@ public static IResourceBuilder<T> WithDockerfile<T>(this IResourceBuilder<T> bui
+        var imageTag = ImageNameGenerator.GenerateImageTag(builder);
+        var annotation = new DockerfileBuildAnnotation(fullyQualifiedContextPath, tempDockerfilePath, stage)
+        {
+            DockerfileFactory = _ => Task.FromResult(dockerfileFactory())
+        };
+
+        return builder.WithAnnotation(annotation, ResourceAnnotationMutationBehavior.Replace)
+                      .WithImageRegistry(registry: null)
+                      .WithImage(imageName)
+                      .WithImageTag(imageTag);
+    }
+
+    /// <summary>
+    /// Causes .NET Aspire to build the specified container image from a Dockerfile generated by an asynchronous factory function.
+    /// </summary>
+    /// <typeparam name="T">Type parameter specifying any type derived from <see cref="ContainerResource"/>.</typeparam>
+    /// <param name="builder">The <see cref="IResourceBuilder{T}"/>.</param>
+    /// <param name="contextPath">Path to be used as the context for the container image build.</param>
+    /// <param name="dockerfileFactory">An asynchronous function that returns the Dockerfile content as a string.</param>
+    /// <param name="stage">The stage representing the image to be published in a multi-stage Dockerfile.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    /// <remarks>
+    /// <para>
+    /// When this method is called, an annotation is added to the <see cref="ContainerResource"/> that specifies the context path
+    /// and a factory function that generates Dockerfile content. The factory is invoked at build time to produce the Dockerfile,
+    /// which is then written to a temporary file and used by the orchestrator to build the container image.
+    /// </para>
+    /// <para>
+    /// The <paramref name="contextPath"/> is relative to the AppHost directory unless it is fully qualified.
+    /// </para>
+    /// <para>
+    /// The factory function is invoked once during the build process to generate the Dockerfile content.
+    /// The output is trusted and not validated.
+    /// </para>
+    /// <example>
+    /// Creates a container called <c>mycontainer</c> with a dynamically generated Dockerfile.
+    /// <code language="csharp">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    ///
+    /// builder.AddContainer("mycontainer", "myimage")
+    ///        .WithDockerfile("path/to/context", async (ct) => 
+    ///        {
+    ///            var template = await File.ReadAllTextAsync("template.dockerfile", ct);
+    ///            return template.Replace("{{VERSION}}", "1.0");
+    ///        });
+    ///
+    /// builder.Build().Run();
+    /// </code>
+    /// </example>
+    /// </remarks>
+    public static IResourceBuilder<T> WithDockerfile<T>(this IResourceBuilder<T> builder, string contextPath, Func<CancellationToken, Task<string>> dockerfileFactory, string? stage = null) where T : ContainerResource
```

Comment:
`Func<CancellationToken, Task<string>> dockerfileFactory` should take the context:

```
Func< DockerfileFactoryContext , Task<string>> dockerfileFactory
```